src/utils/parsers.ts

```diff
@@ -65,6 +66,19 @@ function parseCoreCount(v: string): number | undefined {
   }
 }
 
+function parseSecretName(secretPath: string): [string, string?] {
+  // E.g. If 'vault1:a/b/c', ['vault1', 'a/b/c'] is returned
+  //      If 'vault1', ['vault1, undefined] is returned
+  if (!secretPathNameRegex.test(secretPath)) {
+    throw new commander.InvalidArgumentError(
+      `${secretPath} is not of the format <vaultName>[:<directoryPath>]`,
+    );
+  }
+  // Returns [vaultName, secretName?]
+  const match = secretPath.match(secretPathNameRegex)!;
+  return [match[1], match[2] || undefined];
+}
+
```

Comment:
Why do we have a whole new parser for this? Does the existing not cover it? What change needed to be made that needs this?

Reply:
Existing parsers require a pattern of `vaultName:secretPath`, but a requirement for this PR is that it should list the contents of the entire vault when `polykey secrets list vaultName` is run, which would otherwise raise an error during parsing.

Reply:
I think one can unify the parsers, if the parser always returns `[vaultName, secretPath]`. And all other users of this can just throw the relevant exception...?

There must be examples of this compositionality elsewhere.

Or if the parsers actually just throw the exception directly, then it would be best to name each parser appropriately. Like `parseX` and `parseXY`.

Reply:
> There must be examples of this compositionality elsewhere.

I asked [ChatGPT](https://chatgpt.com/share/e/1ac30d0d-56f7-4f4e-a8d5-4a6585c38344) about this, and it did come up with a way to compose complex parsers based on simple parser "modules". (see https://github.com/MatrixAI/Polykey-CLI/issues/251#issuecomment-2311481854).

Basically, we can define parsers which do simple things like only validate the `vaultName`, or the `secretPath`, and we can compose a new parser on the fly with that. Moreover, an option for adding optional parsers also exist. For example, in the case of `ls` command, the parser can look like this:

```ts
// composeParser(mandatoryParsers, optionalParsers)
const parser = composeParser([vaultName], [secretPath]);

// Run parsing using the parser
// let result = Parse(parser);

// result = {
//   vaultName: 'vaultName'
// }
// --- or ---
// result = {
//   vaultName: 'vaultName',
//   secretPath: 'pathToSecret',
// }

Reply:
There's no need to get that complex. We can just modify the existing parser to allow for just the vaultname. So parsing `'vaultName:'` would return `['vaultName', '',]`. `'vaultName:path'` gives us `['vaultName', 'path']` as usual. But now we allow just `'vaultName'` to return `['vaultname', undefined]`

Wherever we use the existing parser we need to add logic to check that the path is defined if it is required. Actually, we can create a 2nd parser to wrap the changed one so that it enforces the path requirement. We'd have to do this to ensure that the parsing error is handled properly by commander. So...

1. Modify the existing parser to allow parsing just the vault name without the `:` delimiter or the path being provided. Without the delimiter the path will be undefined.
2. creating a 2nd parser function that wraps the first just to enforce requiring the path.
3. Modify existing usage of the parser to use the newly created 2nd parser.

Reply:
I've done that. However, I'm also trying to look for a solution for #251.

Maybe I'll write one which returns only the vault name, which would also solve that issue. As I am already working on simplifying the parsers, I don't see a reason why to add that small change in a whole new PR.